feat(ktor-starter): add Koin DI integration

fireboy1919 opened the pull request:
## Summary

- Refactor serve module as standalone publishable artifact
- Add micronaut-starter demo app with serve-only DI support
- Rename ViaductServerProvider to ViaductProvider
- Add Koin dependency injection integration to ktor-starter demo app
- Implement KoinTenantCodeInjector for resolver instantiation via Koin

**PUBLIC**
feat(ktor-starter): add Koin DI integration with custom TenantCodeInjector

Implements Koin-based dependency injection for the ktor-starter demo app, demonstrating how to use a DI framework as the source of truth for resolver instantiation.

Key components:
- Serve module extracted as standalone publishable artifact
- Micronaut-starter demo app showing DI integration pattern
- KoinTenantCodeInjector: implements TenantCodeInjector, delegates to Koin
- viaductModule: registers Viaduct singleton configured with Koin-based DI
- resolversModule: registers services and resolver factories
- AsciiArtService/AsciiArtResolver: example of constructor injection

## Test Plan

- [x] All existing tests pass
- [x] New ASCII art query test via Koin DI passes
- [x] `./gradlew build` succeeds for ktor-starter
- [x] `./gradlew build` succeeds for micronaut-starter

Commit:
feat(demoapps): add micronaut-starter with serve-only DI support

Add a new demo application demonstrating Micronaut integration with
Viaduct's serve module. Key features:

- MicronautServerProvider: Starts only Micronaut's ApplicationContext
  (DI container) without the HTTP server for fast development iteration
- MicronautTenantCodeInjector: Bridges Micronaut DI with Viaduct resolvers
- ViaductConfiguration: Factory for creating Viaduct bean with DI support

This enables fast startup times during development by avoiding full
Micronaut HTTP server initialization while still providing full
dependency injection for resolvers.

Structure:
- serve/MicronautServerProvider.kt - ViaductServer integration
- injector/ViaductConfiguration.kt - Viaduct bean factory
- injector/MicronautTenantCodeInjector.kt - DI bridge
- resolvers/ - Sample resolver implementations

Commit:
refactor(micronaut-starter): use limited package scanning for faster startup

Use ApplicationContext.builder().packages() to limit Micronaut's
classpath scanning to only the necessary packages:
- com.example.viadapp.injector (ViaductConfiguration, TenantCodeInjector)
- com.example.viadapp.resolvers (Resolver implementations)

This avoids loading controllers, filters, HTTP server components, and
other beans that aren't needed for serve mode, resulting in faster
startup times.

Commit:
feat(ktor-starter): add Koin DI integration with custom TenantCodeInjector

Implements Koin-based dependency injection for the ktor-starter demo app,
demonstrating how to use a DI framework as the source of truth for resolver
discovery and instantiation (analogous to Spring in starwars demo).

Key components:
- KoinTenantCodeInjector: implements TenantCodeInjector, delegates to Koin
- KoinTenantResolverClassFinder: returns resolvers from explicit set (no scanning)
- KoinTenantResolverClassFinderFactory: creates finders for packages
- viaductModule: registers Viaduct singleton configured with Koin-based DI
- resolversModule: registers services and resolver factories
- resolverClasses set: explicit registry for resolver discovery

Example implementation:
- AsciiArtService: loads ASCII art from resources
- AsciiArtResolver: uses constructor injection via Koin
- asciiArt query in GraphQL schema

To add a new resolver:
1. Add class to resolverClasses set
2. Register as factory in resolversModule

Commit:
refactor(ktor-starter): simplify Koin integration

Remove KoinTenantResolverClassFinder and KoinTenantResolverClassFinderFactory.
Use Viaduct's default classpath scanning for resolver discovery.

Koin is now only used for instantiation via KoinTenantCodeInjector,
which enables constructor injection of dependencies into resolvers.

Commit:
refactor(ktor-starter): properly inject Koin dependencies

Register Koin instance itself in the module so it can be injected.
Use get() for all dependency injection rather than explicit getKoin() calls.

Commit:
refactor(serve): extract serve module as standalone publishable artifact

Move serve runtime code from gradle-plugins/application-plugin to a
standalone serve/ module that publishes com.airbnb.viaduct:serve artifact.

This enables micronaut-starter and other apps to depend on the serve
runtime classes (ViaductServerProvider, ViaductServerConfiguration)
when building in isolation with mavenLocal.

- Create serve/ module with viaduct-publishing convention
- Move serve source and test files from application-plugin
- Update application-plugin to depend on libs.viaduct.serve
- Add serve module to included-builds/core/settings.gradle.kts
- Add dependency substitution in root settings.gradle.kts
- Update micronaut-starter to use mavenLocal and correct version

Commit:
refactor(serve): rename ViaductServerProvider to ViaductProvider

- Rename ViaductServerProvider interface to ViaductProvider
- Rename MicronautServerProvider to MicronautViaductProvider
- Reorganize micronaut-starter with dev/production source sets:
  - Production code in src/main/kotlin/.../production/
  - Dev-only code in src/dev/kotlin/.../dev/
  - viaduct-serve dependency only included in dev builds
- Update version to 0.16.0-SNAPSHOT
- Add micronaut-starter to version sync scripts and tests
- Update documentation with new structure and build configuration

rstata commented on demoapps/ktor-starter/resolvers/src/main/resources/ascii/viaduct.txt:
```diff
@@ -0,0 +1,17 @@
+ |  ||    ||    ||    ||    ||    ||    ||    ||  |
+ |  ||    ||    ||    ||    ||    ||    ||    ||  |
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+         GraphQL Data Pipeline Framework
```
I don't know if i like "Data Pipeline Framework" -- how about "Data-Oriented Service Mesh"?

rstata commented on demoapps/ktor-starter/src/test/kotlin/com/example/viadapp/HelloWorldTest.kt:
```diff
@@ -215,4 +215,31 @@ class HelloWorldTest {
+            response.status shouldBe HttpStatusCode.OK
+            // Verify we get ASCII art data (contains GraphQL framework text)
+            val body = response.bodyAsText()
+            body.contains("GraphQL Data Pipeline Framework") shouldBe true
```
change here too if you change it.

Commit:
chore(ktor-starter): update branding and remove mavenLocal

- Change tagline to "Data-Oriented Service Mesh"
- Remove mavenLocal() from repository configurations

Commit:
fix(micronaut-starter): update viaductVersion to match VERSION file

Commit:
fix(ktor-starter): resolve JVM signature clash in ViaductProvider

Remove property-based injection that caused JVM signature clash between
the generated getter and the interface method getViaduct().
